File: apps/api/app/schemas/tracking.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel

from app.models.user_supplement import TakeWindow


class TrackingEventResponse(BaseModel):
    item_id: str
    item_name: str
    item_type: Literal["supplement", "medication", "therapy"]
    take_window: TakeWindow | None
    status: Literal["taken", "skipped"]
    scheduled_at: datetime
    taken_at: datetime | None
    skip_reason: str | None
    regimes: list[str]


class TrackingSuggestionResponse(BaseModel):
    item_id: str | None
    item_name: str | None
    item_type: Literal["supplement", "medication", "therapy", "overall"]
    headline: str
    recommendation: str


class TrackingItemStatResponse(BaseModel):
    item_id: str
    item_name: str
    item_type: Literal["supplement", "medication", "therapy"]
    take_window: TakeWindow
    regimes: list[str]
    scheduled_count: int
    taken_count: int
    skipped_count: int
    pending_count: int
    completion_rate: float
    last_taken_at: datetime | None


class TrackingOverviewResponse(BaseModel):
    window_days: int
    start_date: date
    end_date: date
    item_type_filter: Literal["supplement", "medication", "therapy"] | None
    scheduled_count: int
    taken_count: int
    skipped_count: int
    pending_count: int
    completion_rate: float
    current_streak_days: int
    item_stats: list[TrackingItemStatResponse]
    recent_events: list[TrackingEventResponse]
    suggestions: list[TrackingSuggestionResponse]

File: packages/domain/src/tracking.ts
import type { TakeWindow } from "./user-items";

export interface TrackingOverview {
  window_days: number;
  start_date: string;
  end_date: string;
  item_type_filter: "supplement" | "medication" | "therapy" | "peptide" | null;
  scheduled_count: number;
  taken_count: number;
  skipped_count: number;
  pending_count: number;
  completion_rate: number;
  current_streak_days: number;
  daily_completion: Record<string, boolean | null>;
  item_stats: TrackingItemStat[];
  recent_events: TrackingEvent[];
  suggestions: TrackingSuggestion[];
}

export interface TrackingItemStat {
  item_id: string;
  item_name: string;
  item_type: "supplement" | "medication" | "therapy" | "peptide";
  take_window: TakeWindow;
  regimes: string[];
  scheduled_count: number;
  taken_count: number;
  skipped_count: number;
  pending_count: number;
  completion_rate: number;
  last_taken_at: string | null;
}

export interface TrackingEvent {
  item_id: string;
  item_name: string;
  item_type: "supplement" | "medication" | "therapy" | "peptide";
  take_window: TakeWindow | null;
  status: "taken" | "skipped";
  scheduled_at: string;
  taken_at: string | null;
  skip_reason: string | null;
  regimes: string[];
}

export interface TrackingSuggestion {
  item_id: string | null;
  item_name: string | null;
  item_type: "supplement" | "medication" | "therapy" | "peptide" | "overall";
  headline: string;
  recommendation: string;
}

export interface BatchAdherenceItemResult {
  item_id: string;
  item_type: "supplement" | "medication" | "therapy" | "peptide";
  item_name: string;
  status: "taken" | "skipped";
  scheduled_at: string;
  taken_at: string | null;
  skip_reason: string | null;
}

export interface BatchAdherenceResponse {
  protocol_id: string;
  protocol_name: string;
  date: string;
  status: "taken" | "skipped";
  items_marked: BatchAdherenceItemResult[];
  items_not_due: string[];
}
